### PR TITLE
Ioctl

### PR DESCRIPTION
**Pull request by kotylamichal**

### CODE REVIEW

File: sys/dev/misc/efirt/efidev.c [hunk 1]
```diff
@@ -77,16 +77,31 @@ efidev_ioctl(struct dev_ioctl_args *ap)
...
+		esrt = egtioc->ptr;
+
+		kprintf("IOCTL: esrt->fw_resource_count = %d\n", esrt->fw_resource_count);
+
+		egtioc->table_len = sizeof(*esrt)+ (sizeof(struct efi_esrt_entry_v1) * esrt->fw_resource_count);
+		data = kmalloc(egtioc->table_len, M_TEMP, M_WAITOK);
+
+		kprintf("table_len = %ln\n", &egtioc->table_len);
+
+		error = copyout(data, &egtioc->ptr, egtioc->table_len);
```
Comment (kotylamichal) [inline sys/dev/misc/efirt/efidev.c]: @SergiiDmytruk can you look here? it is a more clean PR with some updates. With that modifications, I got `-1` error from copyout. `esrt->fw_resource_count` value is read correctly, the only problem is passing ESRT values to user space. Can you give me some tips on how to fix that?

Comment (SergiiDmytruk) [reply sys/dev/misc/efirt/efidev.c]: If `egtioc->ptr` is NULL or `egtioc->buf_len < egtioc->table_len`, you shouldn't copy anything. And case of `egtioc->ptr ==  NULL` should return "success" (0) as it's a request for table size.

Comment (kotylamichal) [reply sys/dev/misc/efirt/efidev.c]: @SergiiDmytruk thanks, now `ioctl` related topics are more clear for me. Can you look for  my last changes? I use copyout to move ESRT data into `efi_get_table_ioc` structure. That structure was created to share data between efirt module and userspace, correct?

Comment (kotylamichal) [reply sys/dev/misc/efirt/efidev.c]: @SergiiDmytruk I end with copyout error - `EFAULT 14 Bad address`

**Comment by kotylamichal** [comment]
@TomaszAIR my effort was continued here: https://github.com/3mdeb/DragonFlyBSD/pull/6

File: sys/dev/misc/efirt/efidev.c (hunk 1 above)
Comment (SergiiDmytruk) [reply sys/dev/misc/efirt/efidev.c]: I guess `efi_get_table(&egtioc->uuid, &egtioc->ptr);` was the error here, so I assume things work now.

> That structure was created to share data between efirt module and userspace, correct?

Yes, it's address is what `ioctl()` passes to the kernel.

Comment (kotylamichal) [reply sys/dev/misc/efirt/efidev.c]: @SergiiDmytruk sure, here is continuation: https://github.com/3mdeb/DragonFlyBSD/pull/6

**Pull request closed without merging**